File: backend/app/forks.py
"""Fork-point detection: positions where the model was genuinely torn between options."""

from __future__ import annotations

from collections.abc import Sequence

from .entropy import max_entropy
from .schemas import ForkSettings, TokenInfo


def fork_score(token: TokenInfo) -> float:
    """Uncertainty score in [0, 1]: the mean of normalised entropy and (1 - top-2 margin)."""
    if len(token.top) < 2:
        return 0.0
    norm_entropy = min(1.0, token.entropy / max_entropy(len(token.top)))
    return round(0.5 * norm_entropy + 0.5 * (1.0 - token.margin), 6)


def is_fork(token: TokenInfo, settings: ForkSettings) -> bool:
    """A fork needs a plausible runner-up AND either high entropy or a thin top-2 margin.

    Forced tokens are never flagged: that position has already been branched on.
    """
    if token.forced or len(token.top) < 2:
        return False
    if token.top[1].prob < settings.min_alt_prob:
        return False
    return token.entropy >= settings.entropy_threshold or token.margin <= settings.margin_threshold


def mark_forks(tokens: Sequence[TokenInfo], settings: ForkSettings) -> list[TokenInfo]:
    """Annotate tokens in place with ``fork_score`` and ``is_fork``; returns them as a list."""
    for token in tokens:
        token.fork_score = fork_score(token)
        token.is_fork = is_fork(token, settings)
    return list(tokens)


def rank_forks(tokens: Sequence[TokenInfo], start: int = 0) -> list[int]:
    """Indices of fork tokens at or after ``start``, most uncertain first."""
    idx = [i for i in range(start, len(tokens)) if tokens[i].is_fork]
    return sorted(idx, key=lambda i: (-tokens[i].fork_score, i))

File: frontend/src/lib/forks.ts
// Mirrors backend/app/forks.py so fork thresholds can be tuned live in the UI.
import type { ForkSettings, TokenInfo, Tree } from './types'

export const maxEntropy = (k: number) => Math.log2(Math.max(k, 1) + 1)

export function forkScore(t: TokenInfo): number {
  if (t.top.length < 2) return 0
  const norm = Math.min(1, t.entropy / maxEntropy(t.top.length))
  return 0.5 * norm + 0.5 * (1 - t.margin)
}

export function isFork(t: TokenInfo, s: ForkSettings): boolean {
  if (t.forced || t.top.length < 2) return false
  if (t.top[1].prob < s.min_alt_prob) return false
  return t.entropy >= s.entropy_threshold || t.margin <= s.margin_threshold
}

/** Return a copy of the tree with fork flags recomputed for new thresholds. */
export function remarkTree(tree: Tree, s: ForkSettings): Tree {
  const nodes = Object.fromEntries(
    Object.entries(tree.nodes).map(([id, node]) => [
      id,
      {
        ...node,
        tokens: node.tokens.map((t) => ({ ...t, fork_score: forkScore(t), is_fork: isFork(t, s) })),
      },
    ]),
  )
  return { ...tree, fork_settings: s, nodes }
}
